Improve server shutdown behaviour

RagnarW opened the pull request:
Use same shutdown procedure in RaftServer and CatchupServer.

ServerShutdown now handles errors and throws after both channel and
event loop group have been closed. It will also expose possible failure
when shutting down event loop group.

Also fix logging when unable to shutdwon event loop group within given
time

RagnarW committed:
Improve server shutdown behaviour

Use same shutdown procedure in RaftServer and CatchupServer.

ServerShutdown now handles errors and throws after both channel and
event loop group have been closed. It will also expose possible failure
when shutting down event loop group.

Also fix logging when unable to shutdwon event loop group within given
time

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private interface ErrorRunner
```
Perhaps just use Runnable?

RagnarW replied:
Runnable will not work since it needs to throw.

andrewkerr9000 replied:
We have `ThrowingAction` which does the same but has a type parameter for the throwable type (although it extends Exception)

RagnarW replied:
I would have to wrap the cause of the future since it is a throwable. But if this is to be a private sub interface I don't really see the point.

andrewkerr9000 replied:
:+1:

andrewkerr9000 commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServer.java:
```diff
@@ -208,28 +209,15 @@ private ChannelInboundHandler decoders( CatchupServerProtocol protocol )
         {
-            channel.close().sync();
-            channel = null;
+            new ServerShutdown( log, workerGroup, channel ).shutdown();
```
Maybe use a static method, e.g. `ServerShutdown.shutdown( log, workerGroup, channel );`?

RagnarW replied:
I agree

lutovich commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java:
```diff
@@ -0,0 +1,110 @@
+    {
+        try ( ErrorHandler errorHandler = new ErrorHandler() )
+        {
+            if ( channel != null )
```
Why do we need to both close channel and shutdown the event loop group? Isn't it enough to just do the later? Or channel comes from a different group/bootstrap?

RagnarW replied:
Not sure. I just kept the old logic. It should be enough according to https://netty.io/wiki/user-guide-for-4.x.html#shutting-down-your-application

lutovich commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java:
```diff
@@ -0,0 +1,110 @@
+                }
+                else
+                {
+                    throwable.addSuppressed( t );
```
Super edge case - this will throw if `t == throwable`

lutovich reviewed (commented):
Maybe also add couple tests for `ServerShutdown`.

RagnarW committed:
Simplified shutdown

As mentioned in comments its only needed to close the EventLoopGroup.
This simplifies a lot and no ErrorHandler is needed.

RagnarW committed:
Throw illegal argument exception if EventLoopGroup is null on shutdown

RagnarW committed:
Avoid reusing parameters in CatchupServer

EventLoopGroup is shutdown and created on lifecycle stop and start always

RagnarW committed:
wip

RagnarW committed:
Renaming and adding more tests.

RagnarW committed:
Change handling of unsuccessful bind request

RagnarW committed:
Added Test for real application
